Ghostery throws exception trying to use storageArea.set().

xeenon opened the pull request:
#### 944eaf18a9d63ee99bcbb6ad1fb93f8b3d210cae
<pre>
Ghostery throws exception trying to use storageArea.set().
<a href="https://webkit.org/b/268565">https://webkit.org/b/268565</a>
<a href="https://rdar.apple.com/problem/122118152">rdar://problem/122118152</a>

Reviewed by Jeff Miller.

We need to use JavaScript&apos;s native JSON encoder for storage, since extensions like Ghostery
use toJSON() functions on custom objects to set data in storage.

Teach the code generator how to convert only the top-level keys of a dictionary while
having all the values be JSValue. This allows for the convenience of a dictionary for
the keys, and true values for conversion with JavaScript&apos;s JSON stringifier.

* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::encodeJSONString): Use JSValue&apos;s _toJSONString when it is a JSValue.
(WebKit::encodeJSONData): Ditto.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::anyItemsExceedQuota): Added keyWithError out param for better error logging.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::set): Record JSON/quota errors per key for debugging ease.
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::toNSDictionary): Added ValuePolicy key to stop at top-level.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_platformTypeConstructor): Added support for NSDictionary=StopAtTopLevel attribute.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
(set): Use NSDictionary=StopAtTopLevel for items.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST): Added new test, and also test quota and null.

Canonical link: <a href="https://commits.webkit.org/273949@main">https://commits.webkit.org/273949@main</a>
</pre>

jeffmapple commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
```diff
@@ -387,17 +387,25 @@ size_t storageSizeOf(NSDictionary<NSString *, NSString *> *keysAndValues)
-            *stop = YES;
+            itemExceededQuota = true;
+            keyWithError = key;
+            *stop = true;
```
`*stop` is still a BOOL. 

    *stop = YES;

jeffmapple commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -152,17 +152,28 @@
-    if (anyItemsExceedQuota(serializedData, webExtensionStorageAreaSyncQuotaBytesPerItem)) {
-        *outExceptionString = toErrorString(nil, @"items", @"exceeded maximum size for a single item");
+    if (keyWithError) {
+        *outExceptionString = toErrorString(nil, [NSString stringWithFormat:@"items[`%@`]", keyWithError], @"it is not JSON-serializable");
```
If I'm reading `toErrorString()` correctly, I think the resulting error string would read better without the `it`:

    *outExceptionString = toErrorString(nil, [NSString stringWithFormat:@"items[`%@`]", keyWithError], @"is not JSON-serializable");

xeenon replied:
The full error turns into:

```
Invalid call to storageArea.set(). The 'items[`key`]' value is invalid, because it is not JSON-serializable.
```

So "it" is needed.

jeffmapple commented on Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
```diff
@@ -249,6 +249,12 @@ id toNSObject(JSContextRef context, JSValueRef valueRef, Class containingObjects
         auto *itemValue = toJSValue(context, item);
+        if (valuePolicy == ValuePolicy::StopAtTopLevel) {
+            if (itemValue)
+                [result setObject:itemValue forKey:toNSString(propertyName.get())];
```
Is there a reason we don't use subscripting here?

    result[toNSString(propertyName.get())] = itemValue;

xeenon replied:
No good reason. I can change it.

xeenon committed:
Ghostery throws exception trying to use storageArea.set().
https://webkit.org/b/268565
rdar://problem/122118152

Reviewed by Jeff Miller.

We need to use JavaScript's native JSON encoder for storage, since extensions like Ghostery
use toJSON() functions on custom objects to set data in storage.

Teach the code generator how to convert only the top-level keys of a dictionary while
having all the values be JSValue. This allows for the convenience of a dictionary for
the keys, and true values for conversion with JavaScript's JSON stringifier.

* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::encodeJSONString): Use JSValue's _toJSONString when it is a JSValue.
(WebKit::encodeJSONData): Ditto.
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::anyItemsExceedQuota): Added keyWithError out param for better error logging.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::set): Record JSON/quota errors per key for debugging ease.
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::toNSDictionary): Added ValuePolicy key to stop at top-level.
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
* Source/WebKit/WebProcess/Extensions/Bindings/Scripts/CodeGeneratorExtensions.pm:
(_platformTypeConstructor): Added support for NSDictionary=StopAtTopLevel attribute.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
(set): Use NSDictionary=StopAtTopLevel for items.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST): Added new test, and also test quota and null.

Canonical link: https://commits.webkit.org/273949@main
